### PR TITLE
Introduce ability to lazy load types

### PR DESCRIPTION
**Pull request by crissi**
Enables lazy loading of, when types are used "correctly" by having the same alias in config and in names.

Will make us not fight the underlying framework and make it just work (FAST). Will of cause need some docs to explain how it should be set up.

I noticed about a 4 times speed up, testing on graphql-laravel-performancetest (did have to change all the type alias to match)

### CODE REVIEW

**Comment by mfn** [comment]
Is this the same as https://github.com/rebing/graphql-laravel/pull/394 but optional?

**Comment by crissi** [comment]
> Is this the same as #394 but optional?

this one requires the types are used "correctly" by having the same alias in config and in names

and it can be turned on if people have the correct type setup (I rather like this than #394)

**Comment by crissi** [comment]
> Is this the same as #394 but optional?

it is a replacement for the other one, so one of them should be closed

**Comment by mfn** [comment]
I'm pretty excited about, thanks for your contribution!

Additionally I did some small style fixes; not necessarily because I prefer them but to keep the PR diff clean. Also some small code clarity improvements.

One thing which bothered me was, because due to the setting being `false` by default and basically only one test "testing for it", how can we ensure everything else works correctly when being enabled?

Long story short, I decided to simply run the unit tests twice; once with lazyload_types=false and once with true. The one test which fails when lazyload_types=true is simply skipped in this case.

Thanks!

**Comment by mfn** [comment]
@crissi I'm thinking of making it default to `true` for v2.

My "gut feeling" is that aliasing is a minority thing and the majority would benefit from having it enabled.

What's your take on it?

**Comment by mfn** [comment]
See https://github.com/rebing/graphql-laravel/pull/414

**Comment by crissi** [comment]
I think that is good idea👍

Få Outlook til iOS<https://aka.ms/o0ukef>
________________________________
Fra: Markus Podar <notifications@github.com>
Sendt: lørdag, juli 20, 2019 10:32 AM
Til: rebing/graphql-laravel
Cc: Christian; Mention
Emne: Re: [rebing/graphql-laravel] Introduce ability to lazy load types (#405)


@crissi<https://github.com/crissi> I'm thinking of making it default totrue for v2.

My "gut feeling" is that aliasing is a minority thing and the majority would benefit from having it enabled.

What's your take on it?

**Comment by mfn** [comment]
@crissi I'm actual planning reverting the default of enabling it.

As was discovered in https://github.com/rebing/graphql-laravel/pull/415 (and https://github.com/rebing/graphql-laravel/pull/417 which was ultimately merged), it does also not work with pagination as it's creates an on demand type who name by design differs from the class; this is inherently incompatible.

I could live with the default as long as only aliasing on its own is affected, but pagination is much more useful in practice and that wouldn't be a good experience no matter how much the docs say about this (technically it's the same "problem" with aliasing).

**Comment by crissi** [comment]
Probably a good idea until we find a solution

Få Outlook til iOS<https://aka.ms/o0ukef>
________________________________
Fra: Markus Podar <notifications@github.com>
Sendt: Tuesday, July 23, 2019 11:16:17 AM
Til: rebing/graphql-laravel <graphql-laravel@noreply.github.com>
Cc: Christian <c2n@live.dk>; Mention <mention@noreply.github.com>
Emne: Re: [rebing/graphql-laravel] Introduce ability to lazy load types (#405)


@crissi<https://github.com/crissi> I'm actual planning reverting the default of enabling it.

As was discovered in #415<https://github.com/rebing/graphql-laravel/pull/415> (and #417<https://github.com/rebing/graphql-laravel/pull/417> which was ultimately merged), it does also not work with pagination as it's creates an on demand type who name by design differs from the class; this is inherently incompatible.

I could live with the default as long as only aliasing on its own is affected, but pagination is much more useful in practice and that wouldn't be a good experience no matter how much the docs say about this (technically it's the same "problem" with aliasing).

**Comment by mfn** [comment]
Default has been set to false (again) => https://github.com/rebing/graphql-laravel/pull/422 😢
